sycl/test-e2e/VirtualFunctions/2/1/1/missing-overrides.cpp

```diff
@@ -0,0 +1,141 @@
+
+class Base {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  virtual void increment(int *) { /* do nothhing */ }
+
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  virtual void multiply(int *) { /* do nothhing */ }
+
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  virtual void substract(int *) { /* do nothhing */ }
+};
+
+class IncrementBy1 : public Base {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  void increment(int *Data) override { *Data += 1; }
+};
+
+class IncrementBy1AndSubstractBy2 : public IncrementBy1 {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  void substract(int *Data) override { *Data -= 2; }
+};
+
+class MultiplyBy2 : public Base {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  void multiply(int *Data) override { *Data *= 2; }
+};
+
+class MultiplyBy2AndIncrementBy8 : public MultiplyBy2 {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  void increment(int *Data) override { *Data += 8; }
+};
+
+class SubstractBy4 : public Base {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  void substract(int *Data) override { *Data -= 4; }
+};
+
+class SubstractBy4AndMultiplyBy4 : public SubstractBy4 {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  void multiply(int *Data) override { *Data *= 4; }
+};
+
+Base *constructAnObject(char *Storage, int Index) {
+  switch (Index) {
+  case 0: {
+    auto *Ret = reinterpret_cast<IncrementBy1 *>(Storage);
+    new (Storage) IncrementBy1;
+    return Ret;
+  }
+  case 1: {
+    auto *Ret = reinterpret_cast<IncrementBy1AndSubstractBy2 *>(Storage);
+    new (Storage) IncrementBy1AndSubstractBy2;
+    return Ret;
+  }
+  case 2: {
+    auto *Ret = reinterpret_cast<MultiplyBy2 *>(Storage);
+    new (Storage) MultiplyBy2;
+    return Ret;
+  }
+  case 3: {
+    auto *Ret = reinterpret_cast<MultiplyBy2AndIncrementBy8 *>(Storage);
+    new (Storage) MultiplyBy2AndIncrementBy8;
+    return Ret;
+  }
+  case 4: {
+    auto *Ret = reinterpret_cast<SubstractBy4 *>(Storage);
+    new (Storage) SubstractBy4;
+    return Ret;
+  }
+  case 5: {
+    auto *Ret = reinterpret_cast<SubstractBy4AndMultiplyBy4 *>(Storage);
+    new (Storage) SubstractBy4AndMultiplyBy4;
+    return Ret;
+  }
+
+  default:
+    return nullptr;
+  }
+}
+
+void applyOp(int *DataPtr, Base *ObjPtr) {
+  ObjPtr->increment(DataPtr);
+  ObjPtr->substract(DataPtr);
+  ObjPtr->multiply(DataPtr);
+}
+
+int main() {
+  constexpr size_t Size =
+      std::max({sizeof(IncrementBy1), sizeof(IncrementBy1AndSubstractBy2),
+                sizeof(MultiplyBy2), sizeof(MultiplyBy2AndIncrementBy8),
+                sizeof(SubstractBy4), sizeof(SubstractBy4AndMultiplyBy4)});
+
+  sycl::buffer<char> ObjStorage(sycl::range{Size});
+  char HostStorage[Size];
```

Comment:
Do we need to set the alignment of these to correspond to some common alignment between the types we may place in it?

Reply:
Good question, I think we should indeed align the storage as well. [basic.align.1](https://eel.is/c++draft/basic.align#1) says:

> Attempting to create an object ([[intro.object]](https://eel.is/c++draft/intro.object)) in storage that does not meet the alignment requirements of the object's type is undefined behavior[.](https://eel.is/c++draft/basic.align#1.sentence-4)

Reply:
Resolved in 04dda014340218f9c2696c404029886309052442